### PR TITLE
[DRAFT] Ao3 6471 bookmark word counts

### PR DESCRIPTION
**Pull request by irrationalpie7**
# Pull Request Checklist

* [x] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [x] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [x] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6471

## Purpose

Add the ability to search, filter, and sort by bookmark word counts.

Similarly to https://github.com/otwcode/otwarchive/pull/4452, we define `public_word_count` and `general_word_count` for series and works, and filter/sort by the appropriate one.

Overview of changes:

- Added :word_count to bookmark search form and :words_from/:words_to to bookmark filters
- Bookmarkable query: Make sure we use public_word_count or general_word_count for sorting and/or filtering, as applicable
- Bookmark query: Use score to sort by word_count
- Series: Update bookmarkable_json to have public/general_word_count
- Work: ditto
- Add a series bookmark to fixtures for convenience

**NOTE**: Sorting and filtering seems to be only partially working? It's working when I go to localhost and poke around, whether I'm searching/filtering a user's bookmarks, a tags bookmarkables, or with the bookmark search form. However, I'm struggling to get some of the tests to work, in particular the feature tests.

## Testing Instructions

See Jira.

## References

These pull requests aren't super recent, but I found #3194, which added word count filtering to work search, and #3252, which overhauled bookmark index pages to arrange by bookmarkable, to be useful for figuring out what I needed to change where. #4559 was also useful for understanding the relationship between bookmark and bookmarkable

## Credit

irrationalpie, they/them (with help/troubleshooting on slack from bilka and sarken)

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -56,7 +56,7 @@ services:
     environment:
       - transport.host=localhost
       - bootstrap.memory_lock=false
-      - "ES_JAVA_OPTS=-Xms1500m -Xmx1500m"
+      - "ES_JAVA_OPTS=-Xms1000m -Xmx1000m"
```
Comment (irrationalpie7) [inline docker-compose.yml:59]: Before this change, elastic search was constantly crashing (possibly either because of my docker settings or my computer not having that much ram?). Capping this at a lower amount helped elastic search better manage its memory usage. Here's a [thread I found useful while investigating](https://discuss.elastic.co/t/elasticsearch-process-ended-by-code-137/345399) + the [official elasticsearch docs on setting the heap size](https://www.elastic.co/docs/reference/elasticsearch/jvm-settings#set-jvm-heap-size)

File: app/views/bookmarks/_search_form.html.erb [hunk 2]
```diff
@@ -40,6 +40,13 @@
       <dd>
         <%= f.text_field :bookmarkable_date %>
       </dd>
+      <dt>
+        <%= f.label :word_count, t("bookmarks.filters.word_count") %>
+        <%= link_to_help "bookmark-search-numerical-help" %>
+      </dt>
+      <dd>
+        <%= f.text_field :word_count %>
+      </dd>
```
Comment (irrationalpie7) [inline app/views/bookmarks/_search_form.html.erb]: Visible here: 
image

File: app/views/bookmarks/_filters.html.erb [hunk 3]
```diff
@@ -101,6 +101,14 @@
...
           </dd>
 
+          <dt class="words">
+            <%= f.label :word_count, t("bookmarks.filters.word_count") %>
+            <%= link_to_help "bookmark-search-numerical-help" %>
+          </dt>
+          <dd class="words">
+            <%= f.text_field :word_count %>
+          </dd>
+
```
Comment (irrationalpie7) [inline app/views/bookmarks/_filters.html.erb]: Visible here:
image

Comment (irrationalpie7) [reply app/views/bookmarks/_filters.html.erb]: This is different from the work filter sidebar, which has "From" and "To" fields, but supporting `words_from` and `words_to` in addition to `word_count` added some additional complexity to the query-building logic (like, what if someone sets from and to values, but also puts something like words>10 in the "search within results" text box?). It felt reasonable to only support a single entry field instead

File: test/fixtures/bookmarks.yml [hunk 4]
```diff
@@ -627,3 +627,16 @@ bookmark_00016:
...
+  created_at: 2009-07-18 00:14:46 Z
+  bookmarkable_id: 5
+  delta: true
+  updated_at: 2009-07-18 00:14:46 Z
+  private: false
+  bookmarker_notes: mixed visibility series yay
+  rec: false
+  id: 47
+  user_id: 
+  hidden_by_admin: false
```
Comment (irrationalpie7) [inline test/fixtures/bookmarks.yml:642]: It turns out we didn't have _any_ series bookmarks in the local data you can view at localhost:3000, so I added one of a mixed-visibility series (can be seen at http://localhost:3000/users/reccer/pseuds/reccer/bookmarks if you have the local dev environment going)

File: app/models/search/bookmark_search_form.rb [hunk 5]
```diff
@@ -141,8 +145,9 @@ def bookmarkable_search_results
 
   def sort_options
     [
-      ['Date Bookmarked', 'created_at'],
-      ['Date Updated', 'bookmarkable_date'],
+      ["Date Bookmarked", "created_at"],
+      ["Date Updated", "bookmarkable_date"],
+      ["Word Count", "word_count"]
```
Comment (irrationalpie7) [inline app/models/search/bookmark_search_form.rb:150]: FYI these don't follow the convention for sentence case--do we want to change it? Or leave as-is?

File: app/models/series.rb [hunk 6]
```diff
@@ -265,6 +259,14 @@ def word_count
     self.works.posted.pluck(:word_count).compact.sum
   end
 
+  def public_word_count
+    self.works.posted.unrestricted.pluck(:word_count).compact.sum
+  end
+
+  def general_word_count
+    self.works.posted.pluck(:word_count).compact.sum
+  end
```
Comment (irrationalpie7) [inline app/models/series.rb:258]: Note: double-check that we're computing this value correctly. I'm not sure where we test visible_word_count, so it's possible we already have some tests? But some edge cases off the top of my head would be series containing:
- draft works
- works with draft chapters
- mystery works (unrevealed)
- anonymous works
- archive-locked works
- works hidden by an admin

(I'm also curious if this value is ever different if the creator of the series is the user looking at the value?)

I call this out for this PR specifically, because if we're about to reindex all series it would be good to make sure we have the right numbers computed. My understanding is that it's comparatively much easier to change the word count display values

File: app/models/work.rb [hunk 7]
```diff
@@ -1220,7 +1220,9 @@ def bookmarkable_json
       pseud_ids: anonymous? || unrevealed? ? nil : pseud_ids,
       user_ids: anonymous? || unrevealed? ? nil : user_ids,
       bookmarkable_type: "Work",
-      bookmarkable_join: { name: "bookmarkable" }
+      bookmarkable_join: { name: "bookmarkable" },
+      public_word_count: word_count,
+      general_word_count: word_count
```
Comment (irrationalpie7) [inline app/models/work.rb:1219]: For symmetry with the series tag PRs, I have added `public_word_count` and `general_word_count` fields. However, I wasn't sure if we should remove `word_count`, as perhaps it's used somewhere already? Alternatively, we could use `word_count` and `general_word_count`--let me know!

File: app/views/bookmarks/_search_form.html.erb (hunk 2 above)
Comment (irrationalpie7) [reply app/views/bookmarks/_search_form.html.erb]: Note: on rereading the issue description, "Word count" is supposed to be above "Work language", so this image is out of date

File: app/views/bookmarks/_filters.html.erb (hunk 3 above)
Comment (irrationalpie7) [reply app/views/bookmarks/_filters.html.erb]: Based on a comment from bilka, this has been updated to match the work filter sidebar
